File: ios/Sources/App/Services/NotificationRouter.swift
import Foundation

/// Bridges a notification tap (handled by `AppDelegate`, outside SwiftUI)
/// into the view hierarchy. `RootTabView` observes `pendingWorkspaceID` and
/// switches to that session's tab when it's set.
@MainActor
final class NotificationRouter: ObservableObject {
    static let shared = NotificationRouter()
    @Published var pendingWorkspaceID: UUID?
    // Set alongside `pendingWorkspaceID` only when a "finished" (not
    // "waiting") notification was tapped — the natural next action after
    // an agent finishes is almost always "show me what changed".
    @Published var pendingDiffWorkspaceID: UUID?
    // Set by App Intents (Shortcuts/Siri) to request a tab switch — a
    // second, simpler channel than `pendingWorkspaceID` since an intent
    // like "open my sessions" has no specific workspace to target.
    @Published var pendingTab: RootTab?

    private init() {}
}

File: android/app/src/main/kotlin/dev/termvault/app/notifications/NotificationRouter.kt
package dev.termvault.app.notifications

import dev.termvault.app.navigation.RootTab
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.UUID

/**
 * Bridges a notification tap (handled by [NotificationActionReceiver] /
 * `MainActivity.onNewIntent`, outside Compose) into the navigation layer.
 * Mirrors iOS `NotificationRouter`.
 */
object NotificationRouter {
    private val _pendingWorkspaceId = MutableStateFlow<UUID?>(null)
    val pendingWorkspaceId: StateFlow<UUID?> = _pendingWorkspaceId.asStateFlow()

    // Set alongside pendingWorkspaceId only for a "finished" (not "waiting")
    // notification tap — the natural next action after an agent finishes is
    // almost always "show me what changed".
    private val _pendingDiffWorkspaceId = MutableStateFlow<UUID?>(null)
    val pendingDiffWorkspaceId: StateFlow<UUID?> = _pendingDiffWorkspaceId.asStateFlow()

    private val _pendingTab = MutableStateFlow<RootTab?>(null)
    val pendingTab: StateFlow<RootTab?> = _pendingTab.asStateFlow()

    fun routeToWorkspace(workspaceId: UUID, showDiff: Boolean) {
        _pendingWorkspaceId.value = workspaceId
        _pendingDiffWorkspaceId.value = if (showDiff) workspaceId else null
    }

    fun routeToTab(tab: RootTab) {
        _pendingTab.value = tab
    }

    fun consumeWorkspace() {
        _pendingWorkspaceId.value = null
        _pendingDiffWorkspaceId.value = null
    }

    fun consumeTab() {
        _pendingTab.value = null
    }
}
